Simple partitioned aggregation (#298)
ClickHouse/pg_clickhouse: doc/pg_clickhouse.md

 the number of rows that must be pulled back into Postgres from 1000 (all of
 them) to just 8, one for each node.

### Partitioned Tables

A PostgreSQL [partitioned table] can mix local partitions with foreign
partitions backed by ClickHouse. A common layout offloads cold data to

> ClickHouse while recent data stays in PostgreSQL:

Reviewer: Mixing cold/hot with old/recent. Pick one.

```pgsql
CREATE TABLE events (id int, ts date, val int, amt float8)
    PARTITION BY RANGE (ts);

-- Cold 2023 data lives on ClickHouse
CREATE FOREIGN TABLE events_cold PARTITION OF events
    FOR VALUES FROM ('2023-01-01') TO ('2024-01-01')
    SERVER ch_svr OPTIONS (table_name 'events');

-- Hot 2024 data stays local
CREATE TABLE events_hot PARTITION OF events
    FOR VALUES FROM ('2024-01-01') TO ('2025-01-01');
```

Relocating data from a local to a foreign partition is left as an exercise

> for the reader.

Reviewer: What? Isn't that the point of this PR?

Author: No, this PR enables query planning once mixed partitioning is set up, but my example stored procedure to offload partitions is left out. This way we can iterate on production workflows before trying to codify anything

Reviewer: Okay, but maybe omit this line, which some will find passive-aggressive.

Author: will add example sql that can be used to docs, but that includes generating CREATE TABLE & it'd be nice if it could take fdw server name using binary driver (& not have some hack where we extract connection string from fdw server), so I want https://github.com/ClickHouse/pg_clickhouse/pull/309 for this

Aggregates spanning both local and foreign partitions need [partitionwise
aggregation], which PostgreSQL disables by default:

```pgsql
SET enable_partitionwise_aggregate = on;
```

> With it, PostgreSQL computes a *partial aggregate* below `Append`, then a

Reviewer: ```suggestion
With `enable_partitionwise_aggregate` enabled, PostgreSQL computes a *partial aggregate* below `Append`, then a
```

*finalize aggregate* above combines those partials into result.
pg_clickhouse pushes the foreign partition's partial down to ClickHouse:

```pgsql
try=# EXPLAIN (VERBOSE, COSTS OFF)
       SELECT count(*), sum(val), min(ts), max(ts) FROM events;
                                      QUERY PLAN
-------------------------------------------------------------------------------------
 Finalize Aggregate
   Output: count(*), sum(val), min(ts), max(ts)
   ->  Append
         ->  Foreign Scan
               Output: (PARTIAL count(*)), (PARTIAL sum(val)), (PARTIAL min(ts)), ...
               Relations: Aggregate on (events_cold events)
               Remote SQL: SELECT count(*), sum(val), min(ts), max(ts) FROM events
         ->  Partial Aggregate
               ->  Seq Scan on events_hot
```

#### When partial aggregates push down

PostgreSQL represents a partial aggregate as a *transition state* that the
finalize step combines across partitions. pg_clickhouse can push a partition's

> partial down only when it can express that as a ClickHouse value:

Reviewer: ```suggestion
partial down only when it can express it as a ClickHouse value:
```

*   **Decomposable aggregates** whose transition state is already the final

>     value push down directly: `count`, `sum`, `min`, `max`, `bool_and`/`every`,

Reviewer: ```suggestion
    value, push down directly: `count`, `sum`, `min`, `max`, `bool_and`/`every`,
```

    `bool_or`, `bit_and`, `bit_or`, and `bit_xor`.
*   **`avg` over integers** pushes its `{count, sum}` state as an array.
*   **`avg`, `var_pop`, `var_samp`, `stddev_pop`, and `stddev_samp` over
    floating point** push their `{N, sum, sum of squared deviations}` state as
    an array.

> `FILTER (WHERE …)` pushes down with these.

Reviewer: ```suggestion
`FILTER (WHERE …)` pushes down with these aggregate functions.
```
